### PR TITLE
Prevent Overflow LRU Cache from Exploding

### PR DESCRIPTION
**Pull request by ethDreamer**
## Issue Addressed

* #4757

In a nutshell, the problem with the [`OverflowLRUCache`](https://github.com/sigp/lighthouse/blob/c7ddf1f0b1cfc19a03fcadc7d5ab13d6d595c69d/beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs#L354) before this PR is that it [caches](https://github.com/sigp/lighthouse/blob/c7ddf1f0b1cfc19a03fcadc7d5ab13d6d595c69d/beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs#L56) an [`AvailabilityPendingExecutedBlock`](https://github.com/sigp/lighthouse/blob/c7ddf1f0b1cfc19a03fcadc7d5ab13d6d595c69d/beacon_node/beacon_chain/src/block_verification_types.rs#L255) for each entry, which [contains the entire `BeaconState`](https://github.com/sigp/lighthouse/blob/c7ddf1f0b1cfc19a03fcadc7d5ab13d6d595c69d/beacon_node/beacon_chain/src/block_verification_types.rs#L294). If we end up with multiple forks and many unavailable blocks, this cache can have many entries, causing it to consume large amounts of memory.

I've addressed this by creating a small `StateLRUCache` which will accept the `AvailabilityPendingExecutedBlock`, move the `BeaconState` to a very limited [LRU cache](https://github.com/ethDreamer/lighthouse/blob/e8ef884c544fe51a3d8ddad94df7e4ea43addaf6/beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs#L55) and return a [`DietAvailablityPendingExecutedBlock`](https://github.com/ethDreamer/lighthouse/blob/e8ef884c544fe51a3d8ddad94df7e4ea43addaf6/beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs#L19) which contains all the same data except the `BeaconState` has been replaced with a root.

If many unavailable blocks are stored at the same time in the `DataAvailabiltyCache`, the excess states will be dropped. If those states are needed later they are recovered by loading the parent `BeaconState` from disk and replaying the block.

### CODE REVIEW

File: beacon_node/beacon_chain/src/data_availability_checker/error.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/data_availability_checker/error.rs:29]: Nice, I've been meaning to do a similar sort of refactor

File: beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs]: I think we may want a hybrid of the two types once tree-states is merged. Although tree-states will allow us to keep a lot more states in memory, I still think the on-disk behaviour of `AvailabilityPendingExecutedBlock` is undesirable: storing full beacon states on disk is too much data to write. Instead, we could mark the state as `ssz(skip_serializing, skip_deserializing)` and when loading from disk, fill it in by loading the parent/etc. Or we could store the state in the DB using `put_state`, and load it by `state_root` when loading from disk (tree-states is better at cleaning up dangling states, so this _should_ be OK).

If we take this approach, then I think we should delete the `ssz_tagged_beacon_state` module now, as it's currently unused, and it won't be necessary.

File: beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs]: This is an expensive call, as it's not using the tree hash cache. We could use `update_tree_hash_cache` instead, or even better, pull the state root from `executed_block.block.block().state_root()`. The block has been consensus-verified, so its state root should match the post-state's root.

File: beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs]: We should use `get_advanced_hot_state` here for two reasons:

1. It handles the case where the parent block is the finalized block, and there are skipped slots between the parent block's slot, and the finalized slot (e.g. the block is from slot 30 and it's finalized at slot 32).
2. It's forwards-compatible with tree-states, which will give the highest-slot state from its cache for this call. The max slot could actually be the slot of the block being  processed (we are happy for the DB to do some of the state advancing for us).

File: beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs]: This looks good. The state roots provided may be incomplete, but that's OK. If the block replayer can't get the state root from the iterator it will build the tree hash cache (which we need anyway).

Actually there's currently a small bug in the block replayer's handling of missing state roots (due to `take_while` being destructive) but I'll fix that in a separate PR to unstable.

File: beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs]: Looks good.

**Review by michaelsproul — CHANGES_REQUESTED** [review]
Nice! This is looking great on the whole. I just had a few minor perf tips and style-tweaks to get in before we merge

File: beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs]: PR: #4803

File: beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs
Comment (ethDreamer) [reply beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs]: Hmmm this suggests an alternative design. Perhaps we get rid of `DietAvailabilityPendingExecutedBlock` and change  [`BlockImportData`](https://github.com/sigp/lighthouse/blob/c7ddf1f0b1cfc19a03fcadc7d5ab13d6d595c69d/beacon_node/beacon_chain/src/block_verification_types.rs#L291) to include an `Option<BeaconState>`. This creates the hybrid type you mentioned. Then when serializing we would always serialize as `None`. We would also probably move the `recover_pending_executed_block()` functionality to a method on `AvailabilityPendingExecutedBlock` instead of the `StateLRUCache`.

Then this PR can behave pretty much the same way, but we can continue to use this as you have described once `tree-states` is merged.

What do you think?

Comment (michaelsproul) [reply beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs]: Yeah I think that sounds good! I was trying to suggest a minimal change atop what you've already implemented, but if you're happy to do a more major re-structure then I think an `Option<BeaconState<E>>` is more future proof. Thanks!

File: beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs
Comment (ethDreamer) [reply beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs]: done!

File: beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs
Comment (ethDreamer) [reply beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs]: think I've got it now - let me know if it's wrong

File: beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs
Comment (ethDreamer) [reply beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs]: Okay so I realized this is a bad design since the `ImportData` should only ever be stripped of the state when inside the `DataAvailabiltyChecker` and designing it so that it could be passed around outside the cache while not containing the state forces us to write code to handle that case (which shouldn't happen) in a ton of places. Better to instead contain this complexity within the cache. I've deleted the `ssz_tagged_beacon_state` function and removed `Encode`, `Decode`, & `Clone` from `ImportData` & `AvailabilityPendingExecutedBlock` since we really shouldn't be doing those things with them (and we no longer are with this change).

Perhaps we should decide in the future how best to handle recovering from disk after tree states.. one really easy option would be to just keep the `StateLRUCache` but increase its size considerably after tree states. Keeping the diet type around and separating the state from the cache entries certainly makes [this call](https://github.com/sigp/lighthouse/blob/840907fcd3e9c1cbe789d308461cfc1454721c56/beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs#L548) much faster..

File: beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs]: The call is good, but we should keep the state root that's returned, i.e. `let (parent_state_root, parent_state) = ...`. This is the state root we should pass to the state root iterator. I would rename the current `parent_state_root` drawn from the block to something like `parent_block_state_root` in order to differentiate them.

File: beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs]: Yeah nice, sounds good. I like the deletions in the latest diff :grin:

File: beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs]: Can remove this once the state root is plumbed through as well

File: beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs
Comment (ethDreamer) [reply beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs]: done!

File: beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs
Comment (ethDreamer) [reply beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs]: done!

File: beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs:183]: if we're explicit about building all caches here, is it possible we're doing too much work? for example if we're not verifying signatures, do we need the pubkey cache? and if we're not crossing an epoch boundary do we need the next epoch's committee cache? I think all the caches should be built on demand, so maybe just removing the explicit building would be better. What do you think @michaelsproul

Comment (michaelsproul) [reply beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs:183]: Yeah, I guess in most cases we will need these caches when we process the next block, but in the case of a reorg or a block on a side chain, we may not. So it might be a bit more resilient if we _don't_ build any caches here (less DoS risk).

In the past we've had some issues with caches not getting auto-built when they're required, but I think we're past that now, and hopefully Hydra helps flush out any cases we've missed.

Comment (ethDreamer) [reply beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs:183]: So I did some messing around and it looks like we only need to build the exit cache & the tree hash cache in order to have equality with the original state. Should I build those or build nothing?

Comment (realbigsean) [reply beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs:183]: >  I guess in most cases we will need these caches when we process the next block

Oh true. Well I guess the scenario this mechanism is designed for is when we have a bunch of heads. So if we don't get a next block on a head it'd still be wasted work, right? Also building caches on-demand as we get each head's "next block" might be better to spread out this work over a longer period of time.

> So I did some messing around and it looks like we only need to build the exit cache & the tree hash cache in order to have equality with the original state. Should I build those or build nothing?

Yea sure, building the ones we know we need makes sense to me

Comment (ethDreamer) [reply beacon_node/beacon_chain/src/data_availability_checker/state_lru_cache.rs:183]: okay fixed

**Review by michaelsproul — APPROVED** [review]

**Pull request merged**
